# Introduccion

Pandas es una herramienta de manipulación y análisis de datos de código abierto, rápida, potente, flexible y fácil de usar.

Construido sobre el lenguaje de programación Python.

Los principales tipos de datos que pueden representarse con pandas son:

1. Datos tabulares con columnas de tipo heterogéneo con etiquetas en columnas y filas.
2. Series temporales



# Transformando DataFrames

## Inspeccion de un Dataframe

Cuando obtienes un nuevo DataFrame para trabajar, lo primero que debes hacer es explorar y ver qué contiene. Hay varios métodos y atributos útiles para esto.

```.head()``` devuelve las primeras filas (la "cabeza" del DataFrame).

```.info()``` muestra información sobre cada una de las columnas, como el tipo de datos y la cantidad de valores que faltan.

```.shape``` devuelve el número de filas y columnas del DataFrame.

```.describe()``` calcula algunas estadísticas de resumen para cada columna.

**homelessness** es un DataFrame que contiene estimaciones de la falta de vivienda en cada estado de EE. UU. en 2018. 

La columna **individual** es el número de personas sin hogar que no forman parte de una familia con niños. La columna **family_members** es el número de personas sin hogar que forman parte de una familia con niños. La columna **state_pop** es la población total del estado.

In [7]:
import pandas as pd

homelessness = pd.read_csv('homelessness.csv',index_col=0)

# Imprime el encabezado de homelessness
homelessness.head()

,region,state,individuals,family_members,state_pop
0,East South Central,Alabama,2570.0,864.0,4887681
1,Pacific,Alaska,1434.0,582.0,735139
2,Mountain,Arizona,7259.0,2606.0,7158024
3,West South Central,Arkansas,2280.0,432.0,3009733
4,Pacific,California,109008.0,20964.0,39461588


In [8]:
# Imprime información sobre homelessness
print(homelessness.info())

<class 'pandas.core.frame.DataFrame'>
Index: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   region          51 non-null     object 
 1   state           51 non-null     object 
 2   individuals     51 non-null     float64
 3   family_members  51 non-null     float64
 4   state_pop       51 non-null     int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 2.4+ KB
None


In [9]:
# Imprime la forma de homelessness
print(homelessness.shape)

(51, 5)


In [10]:
# Imprime una descripción de homelessness
print(homelessness.describe())

         individuals  family_members     state_pop
count      51.000000       51.000000  5.100000e+01
mean     7225.784314     3504.882353  6.405637e+06
std     15991.025083     7805.411811  7.327258e+06
min       434.000000       75.000000  5.776010e+05
25%      1446.500000      592.000000  1.777414e+06
50%      3082.000000     1482.000000  4.461153e+06
75%      6781.500000     3196.000000  7.340946e+06
max    109008.000000    52070.000000  3.946159e+07


**Puedes ver que el número promedio de personas sin hogar en cada estado es de aproximadamente 7226. Exploremos el DataFrame más a fondo.**

## Partes de un DataFrame

Para comprender mejor los objetos de tipo DataFrame es útil saber que constan de tres componentes almacenados como atributos:

```.values```: Una matriz bidimensional de valores NumPy.

```.columns```: Un índice de columnas con los nombres de cada columna.

```.index```: Un índice para filas, ya sea números de fila o nombres de fila.

Por lo general, puedes pensar en los índices como una lista de cadenas o números, aunque el tipo de datos pandas **Index** permite opciones más sofisticadas.

In [11]:
# Importa a pandas usando el alias pd
import pandas as pd
homelessness = pd.read_csv('homelessness.csv')

# Imprime los valores de homelessness
print(homelessness.values)

# Imprime el índice de la columnas de homelessness
print(homelessness.columns)

# Imprime el índice de las filas de homelessness
print(homelessness.index)

[[0 'East South Central' 'Alabama' 2570.0 864.0 4887681]
 [1 'Pacific' 'Alaska' 1434.0 582.0 735139]
 [2 'Mountain' 'Arizona' 7259.0 2606.0 7158024]
 [3 'West South Central' 'Arkansas' 2280.0 432.0 3009733]
 [4 'Pacific' 'California' 109008.0 20964.0 39461588]
 [5 'Mountain' 'Colorado' 7607.0 3250.0 5691287]
 [6 'New England' 'Connecticut' 2280.0 1696.0 3571520]
 [7 'South Atlantic' 'Delaware' 708.0 374.0 965479]
 [8 'South Atlantic' 'District of Columbia' 3770.0 3134.0 701547]
 [9 'South Atlantic' 'Florida' 21443.0 9587.0 21244317]
 [10 'South Atlantic' 'Georgia' 6943.0 2556.0 10511131]
 [11 'Pacific' 'Hawaii' 4131.0 2399.0 1420593]
 [12 'Mountain' 'Idaho' 1297.0 715.0 1750536]
 [13 'East North Central' 'Illinois' 6752.0 3891.0 12723071]
 [14 'East North Central' 'Indiana' 3776.0 1482.0 6695497]
 [15 'West North Central' 'Iowa' 1711.0 1038.0 3148618]
 [16 'West North Central' 'Kansas' 1443.0 773.0 2911359]
 [17 'East South Central' 'Kentucky' 2735.0 953.0 4461153]
 [18 'West South Cen

Los **DataFrames** tienen tres componentes: valores, un índice de columnas y un índice de filas.

## Ordenar Filas

Encontrar fragmentos interesantes en un DataFrame suele ser más fácil si cambiamos el orden de las filas. Puedes ordenar las filas pasando un nombre de columna a ```.sort_values()```.

En los casos en que las filas tengan el mismo valor (esto es común si se ordena en una variable categórica), es posible que desees romper los empates ordenando en otra columna. Puedes ordenar varias columnas de esta manera pasando una lista de nombres de columnas.

- una columna: ```df.sort_values("breed")```
- varias columnas: ```df.sort_values(["breed", "weight_kg"])```

In [13]:
homelessness = pd.read_csv('homelessness.csv', index_col=0)

In [14]:
# Ordena homelessness por individuals en orden ascendente
homelessness_ind = homelessness.sort_values("individuals", ascending=True)

# Imprime las primeras filas
print(homelessness_ind.head())

                region         state  individuals  family_members  state_pop
50            Mountain       Wyoming        434.0           205.0     577601
34  West North Central  North Dakota        467.0            75.0     758080
7       South Atlantic      Delaware        708.0           374.0     965479
39         New England  Rhode Island        747.0           354.0    1058287
45         New England       Vermont        780.0           511.0     624358


In [15]:
# Ordena homelessness por family_members en orden descendente
homelessness_fam = homelessness.sort_values('family_members', ascending=False)

# Imprime las primeras filas
print(homelessness_fam.head())

                region          state  individuals  family_members  state_pop
32        Mid-Atlantic       New York      39827.0         52070.0   19530351
4              Pacific     California     109008.0         20964.0   39461588
21         New England  Massachusetts       6811.0         13257.0    6882635
9       South Atlantic        Florida      21443.0          9587.0   21244317
43  West South Central          Texas      19199.0          6111.0   28628666


In [17]:
# Ordena homelessness por region, luego por family_members descendente
homelessness_reg_fam = homelessness.sort_values(['region','family_members'], 
                       ascending = [True,False])

# Imprime las primeras filas
print(homelessness_reg_fam.head())

                region      state  individuals  family_members  state_pop
13  East North Central   Illinois       6752.0          3891.0   12723071
35  East North Central       Ohio       6929.0          3320.0   11676341
22  East North Central   Michigan       5209.0          3142.0    9984072
49  East North Central  Wisconsin       2740.0          2167.0    5807406
14  East North Central    Indiana       3776.0          1482.0    6695497


## Subconjundo de columnas

Cuando trabajamos con datos, es posible que no necesitemos todas las variables en el set de datos. Los corchetes **([])** se pueden usar para seleccionar solo las columnas que nos interesan en el orden que tenga sentido para ti. Para seleccionar solo **"col_a"** del DataFrame **df**, usa:

```df["col_a"]```

Para seleccionar "col_a" y "col_b" de df, usa:

```df[["col_a", "col_b"]]```

In [18]:
# Selecciona la columna individuals
individuals = homelessness['individuals']

# Imprime el encabezado del resultado.
print(individuals.head())

0      2570.0
1      1434.0
2      7259.0
3      2280.0
4    109008.0
Name: individuals, dtype: float64


In [19]:
# Selecciona las columnas state y family_members
state_fam = homelessness[['state','family_members']]

# Imprime el encabezado del resultado.
print(state_fam.head())

        state  family_members
0     Alabama           864.0
1      Alaska           582.0
2     Arizona          2606.0
3    Arkansas           432.0
4  California         20964.0


In [20]:
# Selecciona solo las columnas individuals y state, en ese orden
ind_state = homelessness[['individuals','state']]

# Imprime el encabezado del resultado.
print(ind_state.head())

   individuals       state
0       2570.0     Alabama
1       1434.0      Alaska
2       7259.0     Arizona
3       2280.0    Arkansas
4     109008.0  California


## Subconjundo de filas

Una gran parte de la ciencia de datos se trata de encontrar qué partes de tu set de datos son interesantes. Una de las técnicas más simples para esto es encontrar un subconjunto de filas que coincidan con algunos criterios. Esto a veces se conoce como filtrado de filas o selección de filas.

Hay muchas maneras de crear subconjuntos de un DataFrame, quizás la más común es usar operadores relacionales para devolver **True** o **False** para cada fila, y luego pasar eso entre corchetes.

```
dogs[dogs["height_cm"] > 60]
dogs[dogs["color"] == "tan"]
```

Puedes filtrar por varias condiciones a la vez usando el operador **&**.

```dogs[(dogs["height_cm"] > 60) & (dogs["color"] == "tan")]```


In [21]:
# Filtra las filas donde individuals es mayor a 10000
ind_gt_10k = homelessness[homelessness['individuals'] > 10000]

# Imprime el resultado
print(ind_gt_10k)

                region       state  individuals  family_members  state_pop
4              Pacific  California     109008.0         20964.0   39461588
9       South Atlantic     Florida      21443.0          9587.0   21244317
32        Mid-Atlantic    New York      39827.0         52070.0   19530351
37             Pacific      Oregon      11139.0          3337.0    4181886
43  West South Central       Texas      19199.0          6111.0   28628666
47             Pacific  Washington      16424.0          5880.0    7523869


In [22]:
# Filtra las filas donde la region es Mountain
mountain_reg = homelessness[homelessness['region'] == 'Mountain']

# Imprime el resultado
print(mountain_reg)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
5   Mountain    Colorado       7607.0          3250.0    5691287
12  Mountain       Idaho       1297.0           715.0    1750536
26  Mountain     Montana        983.0           422.0    1060665
28  Mountain      Nevada       7058.0           486.0    3027341
31  Mountain  New Mexico       1949.0           602.0    2092741
44  Mountain        Utah       1904.0           972.0    3153550
50  Mountain     Wyoming        434.0           205.0     577601


In [23]:
# Filtra las filas donde family_members sea menor a 1000 y la region es Pacific
fam_lt_1k_pac = homelessness[(homelessness['family_members'] < 1000) & 
                            (homelessness['region'] == 'Pacific')]

# Imprime el resultado
print(fam_lt_1k_pac)

    region   state  individuals  family_members  state_pop
1  Pacific  Alaska       1434.0           582.0     735139


## Subconjunto de filas con variables categoricas

La creación de subconjuntos de datos en función de una variable categórica a menudo implica el uso del operador "o" **(|)** para seleccionar filas de varias categorías. Esto puede volverse tedioso cuando buscamos los estados de una de tres regiones diferentes, por ejemplo. En su lugar, utiliza el método **.isin()** , que nos permitirá abordar este problema escribiendo una sola condición en lugar de tres separadas.

```
colors = ["brown", "black", "tan"]
condition = dogs["color"].isin(colors)
dogs[condition]
```

In [24]:
# Selecciona las filas en la region South Atlantic o Mid-Atlantic
regions = ['South Atlantic','Mid-Atlantic']
condition = homelessness['region'].isin(regions)
south_mid_atlantic = homelessness[condition]

# Imprime el resultado
print(south_mid_atlantic)

            region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9 

In [25]:
# Los estados del desierto de Mojave
canu = ["California", "Arizona", "Nevada", "Utah"]

# Filtra las filas en los estados del desierto de Mojave
condition = homelessness['state'].isin(canu)
mojave_homelessness = homelessness[condition]

# Imprime el resultado
print(mojave_homelessness)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
4    Pacific  California     109008.0         20964.0   39461588
28  Mountain      Nevada       7058.0           486.0    3027341
44  Mountain        Utah       1904.0           972.0    3153550


## Agregando nuevas columnas

No estás limitado solo a los datos que te dan. Siempre puedes agregar nuevas columnas a un DataFrame. Esto tiene muchos nombres, como transformación, mutación e ingeniería de características.

Puedes crear nuevas columnas desde cero, pero también es común derivarlas de otras columnas, por ejemplo, agregando columnas o cambiando sus unidades.

In [27]:
# Agrega la columna total como la suma de individuals y family_members
homelessness['total'] = homelessness['individuals'] + homelessness['family_members']

# Agrega p_individuals como la proporción del total que son individuos
homelessness['p_individuals'] = (homelessness['individuals'] / homelessness['total'])

# Imprime el resultado
print(homelessness.head())

               region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  
0    3434.0       0.748398  
1    2016.0       0.711310  
2    9865.0       0.735834  
3    2712.0       0.840708  
4  129972.0       0.838704  


## Uso de todos los conceptos anteriores

Los cuatro tipos de datos más comunes de manipulación de datos: ordenar filas, creación de subconjuntos de columnas, creación de subconjuntos de filas y adición de nuevas columnas. En un análisis de datos en la vida real, puedes mezclar y combinar estas cuatro manipulaciones para responder a una multitud de preguntas.

En este ejercicio, se respondera la pregunta "¿Qué estado tiene el mayor número de personas sin hogar por cada 10,000 personas?"

In [29]:
# Crea la columna indiv_per_10k como individuos sin hogar por 10k pop estatal
homelessness["indiv_per_10k"] = 10000 * homelessness['individuals'] / homelessness['state_pop'] 

# Crea el subconjunto de filas donde indiv_per_10k sea mayor que 20
high_homelessness = homelessness[homelessness['indiv_per_10k'] > 20]

# Ordena high_homelessness de forma descendente por indiv_per_10k
high_homelessness_srt = high_homelessness.sort_values('indiv_per_10k',ascending=False)

# Usando high_homelessness_srt, selecciona las columnas state e indiv_per_10k
result = high_homelessness_srt[['state','indiv_per_10k']]

# Imprime result
print(result)

                   state  indiv_per_10k
8   District of Columbia      53.738381
11                Hawaii      29.079406
4             California      27.623825
37                Oregon      26.636307
28                Nevada      23.314189
47            Washington      21.829195
32              New York      20.392363


**El Distrito de Columbia tiene el mayor número de personas sin hogar: casi 54 por cada diez mil personas.** 

Esto es casi el doble del número del siguiente estado más alto, **Hawaii**. Si combinas la adición de nuevas columnas, subconjuntos de filas, clasificación y selección de columnas, puedes responder muchas preguntas como esta.

# Agregando DataFrames

## Media y mediana

Media y mediana
Las estadísticas de resumen son exactamente lo que su nombre implica - resumen muchos números en una estadística. Por ejemplo, media, mediana, mínimo, máximo y desviación estándar son estadísticas de resumen. Calcular estadísticas de resumen te permite conocer mejor tus datos, incluso si hay muchos.

```sales``` está disponible con datos aleatorios de muestra.

In [3]:
import pandas as pd
sales = pd.read_csv("sales.csv")
sales.head()

,store,type,department,date,weekly_sales,is_holiday,temperature_c,fuel_price_usd_per_l,unemployment
0,27,B,76,2018-01-01,45499.288343,False,36.196223,0.681140,8.589916
1,48,C,88,2018-01-02,26229.673084,True,-7.070981,1.218215,5.026938
2,28,A,87,2018-01-03,45679.722355,True,9.672497,0.689596,8.970767
3,32,C,42,2018-01-04,33759.340710,True,26.962087,1.155827,3.668313
4,68,A,28,2018-01-05,30133.629943,True,-2.462690,0.942440,6.628881


In [4]:
# Imprime el encabezado del DataFrame sales
print(sales.head())

# Imprime la información sobre el DataFrame sales
print(sales.info())

# Imprime la media de weekly_sales
print(sales["weekly_sales"].mean())

# Imprime la mediana de weekly_sales
print(sales["weekly_sales"].median())

   store type  department        date  weekly_sales  is_holiday  \
0     27    B          76  2018-01-01  45499.288343       False   
1     48    C          88  2018-01-02  26229.673084        True   
2     28    A          87  2018-01-03  45679.722355        True   
3     32    C          42  2018-01-04  33759.340710        True   
4     68    A          28  2018-01-05  30133.629943        True   

   temperature_c  fuel_price_usd_per_l  unemployment  
0      36.196223              0.681140      8.589916  
1      -7.070981              1.218215      5.026938  
2       9.672497              0.689596      8.970767  
3      26.962087              1.155827      3.668313  
4      -2.462690              0.942440      6.628881  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10774 entries, 0 to 10773
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   store                 10774 non-null  int64  
 1   t

## Resumiendo fechas

Las estadísticas de resumen también se pueden calcular en columnas de fechas que tienen valores con el tipo de datos **datetime64**. Algunas estadísticas de resumen, como la media, no tienen mucho sentido en las fechas, pero otras son súper útiles, por ejemplo, mínimo y máximo, que te permiten ver qué rango de tiempo cubren tus datos.

In [5]:
# Imprime el máximo de la columna date
print(sales["date"].max())

# Imprime el mínimo de la columna date
print(sales["date"].min())

2047-07-01
2018-01-01


Tomar el mínimo y máximo de una columna de fechas es útil para averiguar qué período de tiempo cubren sus datos

## Resumenes eficientes

Si bien pandas y NumPy tienen muchas funciones, a veces, es posible que necesites una función diferente para resumir tus datos.

El método ```.agg()``` te permite aplicar tus propias funciones personalizadas a un DataFrame, así como aplicar funciones a más de una columna de un DataFrame a la vez, haciendo que tus agregaciones sean super eficientes. Por ejemplo,

```df['column'].agg(function)```

En la función personalizada para este ejercicio, "IQR" es la abreviatura de rango inter-cuartil, que es el percentil 75 menos el percentil 25. Es una alternativa a la desviación estándar que es útil si los datos contienen valores atípicos.

In [6]:
# Una función IQR personalizada
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)
    
# Imprime IQR de la columna temperature_c
print(sales["temperature_c"].agg(iqr))

25.365085975466343


In [7]:
# Una función IQR personalizada
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Actualiza el código para imprimir IQR de temperature_c, fuel_price_usd_per_l y unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg(iqr))

temperature_c           25.365086
fuel_price_usd_per_l     0.503432
unemployment             3.423676
dtype: float64


In [10]:
# Importa NumPy y crea una función IQR personalizada
import numpy as np
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Actualiza el código para imprimir IQR y np.median de temperature_c, fuel_price_usd_per_l, y unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg([iqr,'median','sum']))

        temperature_c  fuel_price_usd_per_l  unemployment
iqr         25.365086              0.503432      3.423676
median      15.050535              1.001134      6.545020
sum     162821.283777          10783.399251  70276.428026


**¡Excelente eficiencia! El método .agg() facilita el cálculo de múltiples estadísticas en varias columnas, todo en una sola línea de código.**

## Estadisticas acumuladas

Las estadísticas acumuladas también pueden ser útiles para dar seguimiento de las estadísticas de resumen a lo largo del tiempo. En este ejercicio, calcularás la suma acumulada y el máximo acumulativo de las ventas semanales de un departamento, lo que te permitirá identificar cuáles fueron las ventas totales hasta ahora, así como cuáles fueron las ventas semanales más altas hasta ahora.

Se ha creado para ti un DataFrame llamado ```sales_1_1```, que contiene los datos de ventas del departamento 1 de la tienda 1.

In [12]:
sales_1_1 = sales[(sales["store"]) == 1 & (sales["department"] == 1)]
sales_1_1.head()

,store,type,department,date,weekly_sales,is_holiday,temperature_c,fuel_price_usd_per_l,unemployment
3905,1,C,1,2028-09-10,15302.432998,False,23.103870,1.081979,3.767605
9751,1,A,1,2044-09-12,10349.284338,False,11.090935,1.082905,3.306716


In [13]:
# Ordena sales_1_1 por date
sales_1_1 = sales_1_1.sort_values(by="date")

# Obtén la suma acumulada de weekly_sales, y añádela como cum_weekly_sales
sales_1_1['cum_weekly_sales'] = sales_1_1['weekly_sales'].cumsum()

# Obtén el máximo acumulativo de weekly_sales y añádelo como cum_max_sales
sales_1_1["cum_max_sales"] = sales_1_1['weekly_sales'].cummax()

# Consulta las columnas que calculaste
print(sales_1_1[["date", "weekly_sales", "cum_weekly_sales", "cum_max_sales"]])

            date  weekly_sales  cum_weekly_sales  cum_max_sales
3905  2028-09-10  15302.432998      15302.432998   15302.432998
9751  2044-09-12  10349.284338      25651.717335   15302.432998


**No todas las funciones que hacen cálculos en columnas devuelven un solo número. Algunas, como las funciones estadísticas acumulativas, devuelven una columna completa.**